web-app/src/common/graphWinPrediction.ts

```diff
@@ -0,0 +1,131 @@
+import * as jkstra from 'jkstra';
+import {
+  doesCellExist, doesCellExistAndHaveStone, GameState, StoneColor,
+  BLACK_NODE_START, BLACK_NODE_END, WHITE_NODE_START, WHITE_NODE_END
+} from './gameState';
+import { Coordinates } from './utils';
+
+export const UNDEFINED_EDGE_COST = -1;
+export const LOW_EDGE_COST = 1;
+export const HIGH_EDGE_COST = 100;
+
+export interface HexBoardWinPredictionGraph {
+  winPrediction: jkstra.Graph;
+  vertexMap: Map<string, jkstra.Vertex>;
+}
+
+/**
+ * Create a graph to represent the winnable paths of a given board.
+ * It doesn't add vertex/edges for the cells of the opposite color because it's not needed for this calculation.
+ */
+export function createWinPredictionGraph(
+  gameState: GameState,
+  stoneColor: StoneColor,
+): HexBoardWinPredictionGraph {
+  const hexBoardWinPredictionGraph: HexBoardWinPredictionGraph = {
+    winPrediction: new jkstra.Graph(),
+    vertexMap: new Map<string, jkstra.Vertex>(),
+  };
+  createVertices(gameState, hexBoardWinPredictionGraph, stoneColor);
+  createStartEndEdges(gameState, hexBoardWinPredictionGraph, stoneColor);
+  createPlayableCellEdges(gameState, hexBoardWinPredictionGraph, stoneColor);
+  return hexBoardWinPredictionGraph;
+}
+
+function createVertices(
+  gameState: GameState,
+  hexBoardWinPredictionGraph: HexBoardWinPredictionGraph,
+  stoneColor: StoneColor
+) {
+  if (stoneColor === "black") {
+    addVertex(hexBoardWinPredictionGraph, BLACK_NODE_START);
+    addVertex(hexBoardWinPredictionGraph, BLACK_NODE_END);
+  }
+  else {
+    addVertex(hexBoardWinPredictionGraph, WHITE_NODE_START);
+    addVertex(hexBoardWinPredictionGraph, WHITE_NODE_END);
+  }
+
+  // Add vertex for each cell on the board that's are not of the opposite color
+  gameState.board.forEach((row, y) => {
+    row.forEach((_cell, x) => {
+      if (_cell.value === stoneColor || _cell.value === "empty") {
+        addVertex(hexBoardWinPredictionGraph, `${y}-${x}`);
+      }
+    });
+  });
+}
+
+function addVertex(hexBoardWinPredictionGraph: HexBoardWinPredictionGraph, id: string) {
+  hexBoardWinPredictionGraph.vertexMap.set(id, hexBoardWinPredictionGraph.winPrediction.addVertex(id));
+}
+
+function createPlayableCellEdges(
+  gameState: GameState,
+  hexBoardWinPredictionGraph: HexBoardWinPredictionGraph,
+  stoneColor: StoneColor,
+) {
+  gameState.board.forEach((row, y) => {
+    row.forEach((cell, x) => {
+      if (cell.value == stoneColor || cell.value == "empty") {
+        const currentCell: Coordinates = { y: y, x: x };
+        const directNeighborCells = [
+          { y: y, x: x - 1 }, { y: y, x: x + 1 },
+          { y: y - 1, x: x }, { y: y + 1, x: x },
+          { y: y - 1, x: x - 1 }, { y: y + 1, x: x + 1 },
+        ];
```

Comment:
I'm a bit surprised by this. I expected something like this:

```suggestion
        const directNeighborCells = [
          { y: y, x: x - 1 }, { y: y, x: x + 1 },
          { y: y - 1, x: x }, { y: y + 1, x: x },
          { y: y - 1, x: x + 1 }, { y: y + 1, x: x - 1 },
        ];
```

I'm gonna run some tests, maybe I'm mistaken...

Reply:
As I feared, I believe you made a mistake here 🙁 .

This test should pass, but it doesn't:

```ts
  it('Should return a valid winning path cost hum...', () => {
    const input = parseGameStateFromMultilineString(`
  ⬡ ⬡ W
   ⬡ W ⬡
    W ⬡ ⬡
  `);

    const cost = getWinnablePathCost(input, 'white');

    expect(cost).toStrictEqual(3);
  });
```

It returns a cost of 201 instead of 3

Reply:
Good catch !